### ISSUE TITLE
Thread safety issues

### ISSUE THREAD

**Issue by vemv** [kkinnear/zprint#226]
Hi there again,

I'm finding that a snippet such as the following demonstrates that zprinting isn't idempotent when run in parallel:

```clj
;; Should return (true true true true), but often false values can be seen
(pmap (fn [specimen]
          (let [format (fn [s]
                         (zprint/zprint-file-str s "_" {:color? false
                                                        :width 120,
                                                        :style [:community :justified :map-nl :pair-nl :binding-nl :rod :hiccup],
                                                        :map {:sort? false}}))]
            (= (format specimen)
               (format (format specimen))
               (format (format (format specimen))))))
        [(slurp "a.clj")
         (slurp "b.clj")
         (slurp "c.clj")
         (slurp "d.clj")])
;; Results become correct by simply changing pmap -> map
```

It's as if some internal state was unintentionally shared between invocations.

**Comment by kkinnear** [comment]
Wow, that's a cool test.  Very concise!

When I first read this issue, my first thought was to go on and on about how zprint certainly has internal state in the way that it handles both strings and structures, and it has one formatting engine and maps a broad set of functions between operating on zippers (produced when formatting code and parsing with `rewrite-clj`) and structures, where no parsing is required.  And while that is all true, I don't think that is what is happening here.  First, because you are using `zprint-file-str`, which is always parsing things. 

I think the problem here is that `zprint-file-str` internally actually uses `set-options!`, and while it puts them all back when it is finished (or at least I think it does), there could well be issues with running multiple `zprint-file-str` invocations simultaneously.

If I can reproduce your test failure, I think I can try some things that should allow me to determine if it is the `set-options!` calls that are the problem.  If so, I expect I could fix that, though it will require a bit of thought as to exactly how that might work.  But first I need to see if that is, indeed, the problem.

In passing, let me point out that zprint (and `zprint-file-str`) uses `pmap` as well by default.  I don't think that is what is causing this issue, but it is certainly an assist to performance.  I seem to remember an improvement of about 20% on a reasonably large file.  You can configure parallel or non-parallel operation in zprint by changing the `:parallel` key in the options map.  I mention this not because I think it will affect this particular issue, but it might mean that you didn't feel the need to run multiple invocations of `zprint-file-str` in parallel if you knew that zprint was already parallel (at least to some extent) internally.

**Comment by kkinnear** [comment]
Great test!  I can reproduce it quite well.  Indeed, I've also found a file that never yields true, whether `map` or `pmap`!

**Comment by kkinnear** [comment]
My theory, that it was the use of `set-options!` inside of `zprint-file-str` seems to be holding up to testing in the real world.  That is, when I remove the options map from `zprint-file-str` and just give it `{}`, I don't get any false values.  If I have that options map on `zprint-file-str` I get some `false` values about every other run, but I can run 20 times and not have any `false` values if I have a `{}` options map on the `zprint-file-str` call.

I then took the options map you have on the `zprint-file-str`, and manually did a `set-options!` with it, and then ran the test with a `{}` in `zprint-file-str`, and it continues to work without any `false` values.  So that is one workaround -- set the options yourself, don't do it on the `Zprint-file-str` call.

I'll spend some more time looking into `zprint-file-str`, and see what it will take to make it not conflict with itself, but hopefully that workaround will help if this was blocking whatever you are doing.

**Comment by vemv** [comment]
> You can configure parallel or non-parallel operation in zprint by changing the :parallel? key in the options map.

Thanks for pointing out this! I'll prefer to set it to false and take care of parallelism from my side (normally I format a few files at a time, mapping 1 thread per file might be more efficient or at least more stable)

> I'll spend some more time looking into zprint-file-str, and see what it will take to make it not conflict with itself, but hopefully that workaround will help if this was blocking whatever you are doing.

Hmm. Anyway, not sure if `zprint-file-str` was the right fn to call, to begin with. I was a bit confused by all the offered options. I essentially want a defn that reformats either a single file, or a file's contents expressed as a string.

If there's a defn that is more idiomatic for that, awesome. Else there's no issue in waiting for a fix and abstaining from parallelism in the meantime.

**Comment by kkinnear** [comment]
No, `zprint-file-str` is what you want -- it will do a file as a string.  That is what the compiled binary programs use, and that is what it was made to do.  It is also the only thing that implements the comment API (where you can put `;!zprint {<options-map>}` in the code to adjust the formatting for one or more functions).

Overall, you will indeed get much better parallelism if you run `zprint-file-str` in parallel with itself then if you just rely on internal parallelism within it (not that the internal parallelism won't help too).  If you are doing multiple files with the same options map, you can just set the options with `set-options!` and then do the parallel thing with the multiple files.  If you are doing multiple files with __different__ options maps on each `zprint-file-str` call, you will need to hold off on the parallelism until I figure out how to fix it to work in parallel with a non-empty option map.

**Comment by vemv** [comment]
Ah, got it!

I understand the workaround now, will apply it in the meantime. Thanks!

**Comment by kkinnear** [comment]
Sigh. I spoke a bit too soon.  The workaround will work as long as nobody uses the comment API I mentioned above.  If they do, parallel calls of `zprint-file-str` will not work at all, since (most) every comment API line does a `set-options!` with the values.  `zprint-file-str` is still the right routine to use.  

I would avoid parallel `zprint-file-str` calls (at least in released code) until I fix this problem.    Which I will work on and figure out before I release `1.2.3`, which is coming pretty soon, I think.  I was thinking maybe this week or next, but I want to get this fixed too, so it might be another week or so after that.

**Comment by vemv** [comment]
Alright! 🍻

**Comment by kkinnear** [comment]
This is fixed in `1.2.3`.  Thanks again for submitting the issue!

**Comment by vemv** [comment]
Awesome! Will try it out. Assuming it's fixed anyway
